### PR TITLE
🐛 [#430][d] - Prevent concurrent stock overselling and enforce balance invariants 🔒

### LINKED ISSUE #430: 🐛 Prevent concurrent stock overselling and enforce balance invariants
## Description

Make every Stock mutation safe under concurrent requests and enforce the balance rules at application and database boundaries.

## Reason

StockOut currently reads availability and decrements later without a row lock. Two requests can approve the same balance and oversell. Receiving can also race while creating the first Location+Variant row.

## Objective

Guarantee atomic Stock creation/mutation and explicit nonnegative on-hand/reserved invariants for existing and future receiving, reservation and transfer flows.

## ✅ Technical Tasks

- [x] Define the allowed balance policy, including whether explicit adjustments may cross zero.
- [x] Use row locks or equivalent atomic mutation for stock-out and reusable Stock mutation services.
- [x] Handle concurrent first-row creation under the unique Location+Variant constraint.
- [x] Enforce on_hand ≥ 0, reserved ≥ 0 and reserved ≤ on_hand at the appropriate database/application layers.
- [x] Keep movement creation and balance mutation in the same transaction.
- [x] Add deterministic concurrent-request and rollback regression tests.

## 🎯 Acceptance Criteria

- [x] Two concurrent outbound requests cannot consume the same available units.
- [x] Concurrent receipt/create paths cannot duplicate or lose a Stock row.
- [x] Invalid balances are rejected even outside HTTP FormRequest validation.
- [x] A failed movement leaves both Stock and movement history unchanged.

## 🔗 References

- Current StockOutService and OpeningBalanceService
- Authorization correction: #400

### 📊 Estimates
- **Optimistic:** `4h` · **Pessimistic:** `8h` · **Tracked:** `6h55m`

### 📅 Sessions
```json
[
  { "date": "2026-08-12", "start": "20:23", "end": "23:34" },
  { "date": "2026-08-13", "start": "16:54", "end": "18:32" },
  { "date": "2026-08-13", "start": "20:55", "end": "23:01" }
]
```

## 📊 Retrospective

**Actual total:** 6h55m across three sessions (2026-08-12 20:23–23:34; 2026-08-13 16:54–18:32;
2026-08-13 20:55–23:01)
- Session 1: 3h11m — context/plan, TDD implementation (migration, `StockMutationService`,
  `Stock` model guards, `StockOutService`/`OpeningBalanceService` refactor, full test suite),
  docs/checklist, PR open, CI gate, Copilot review response cycle, commit squash, Devin/DeepWiki
  review response cycle, close-out.
- Session 2: 1h38m — a further public review pass (`chatgpt-codex-connector`) on the still-open
  PR flagged that the `NOT VALID` balance constraints added in Session 1 were never repaired or
  validated against pre-existing rows, so the advertised invariant wasn't actually guaranteed for
  data already in the table. Fixed with a clamping `UPDATE` plus `VALIDATE CONSTRAINT` for all
  three constraints, verified with the targeted test suite and Pint, then replied to and resolved
  the thread.
- Session 3: 2h6m — final PR close-out: pre-flight re-validation (review threads, mergeable
  state), branch squash, and this issue's own finalization/retrospective update.

**Variance vs. estimate:** 2h55m over the 4h optimistic estimate, 1h5m under the 8h pessimistic
estimate.

**Narrative:** Landed under the pessimistic estimate but over the optimistic one, driven mainly by
a third review cycle that arrived after the issue had already been finalized once (Session 1). The
implementation itself (DB CHECK constraints, `StockMutationService`, guarded `Stock` model methods,
refactoring both `StockOutService` and `OpeningBalanceService` to route through it, plus 45
new/updated tests) matched the estimate's assumption of a contained, well-scoped concurrency fix —
no reservation/transfer flow existed yet to build around, which kept the diff to the two existing
services plus the shared safety layer. Session 1's two review cycles added real but modest time:
Copilot's first pass found a genuine gap (the `reserved <= on_hand` guard on `decreaseOnHand` only
checked for negative `on_hand`, not for dropping below `reserved`) plus a docblock inaccuracy —
both fixed with a small commit and matching test. Devin's DeepWiki pass then surfaced two
"Investigate"-severity design points worth acting on (the migration would fail on any pre-existing
row already violating the invariant — fixed by adding the CHECK constraints as `NOT VALID`; and
`receiveInto()`'s original optimistic-INSERT-first design made the doomed-write path the *common*
case for routine repeat receipts — restructured to check for an existing row first, moving the
race-recovery logic into a separately-testable `insertOrRecoverFromRace()`), plus a genuine
input-validation gap (non-positive quantities could silently invert a mutation, which the model
guards now reject) and a vacuous test assertion (missing `->fresh()` on a DB-computed column,
mirroring a fix already applied elsewhere in this same PR). All four were fixed with matching
test coverage; the two remaining Informational flags (exception-message detail exposure — matches
an existing codebase-wide pattern; a very narrow `ModelNotFoundException` edge case requiring a
zero-stock row to be deleted in the exact microsecond window of an unrelated receive race) were
evaluated and left as-is with reasoning recorded, not fixed, since neither represents a real
regression or a reachable defect worth the added complexity. Devin's own live re-scan of the final
Session 1 commit did not finish within a reasonable window (free-tier scan appeared to stall) — the
fixes were still applied and tested based on its prior findings, and CI (including the full
1322-test Feature suite) stayed green throughout. What Session 1 could not anticipate was that the
`NOT VALID` fix itself would draw a *further* finding: adding a constraint `NOT VALID` stops future
violations but silently leaves any row that already violates it uncorrected and unvalidated,
meaning the invariant this issue set out to guarantee wasn't actually true for existing data. That
gap — caught by a fourth, independent reviewer after the PR had already gone through two rounds and
been marked complete — is exactly the kind of thing a contained unit of review cycles can miss, and
justifies the extra session rather than reflecting scope creep in the original implementation.

### PR DESCRIPTION
**Pull request by pakodiazdev**
## Summary
Closes #430
Devin Review: https://deepwiki.com/pakodiazdev/sushigo/pull/447

- 🗄️ Added Postgres CHECK constraints on `stock`: `on_hand >= 0`, `reserved >= 0`, `reserved <= on_hand`
- 🆕 Added `StockMutationService`: a reusable, transaction-safe entry point for locked reads (`lockAndGet`) and guarded mutations (`increaseOnHand`, `decreaseOnHand`), shared by stock-out today and by future receiving/reservation/transfer flows
- 🔒 `StockOutService` now locks the `Stock` row for update before checking availability, holding the lock for the rest of the transaction so a second concurrent stock-out can't read a stale `on_hand` and oversell
- ♻️ `OpeningBalanceService` now uses an optimistic-insert-then-recover strategy (`receiveInto`) for the first receipt at a location+variant: a concurrent request racing to create the same row hits the unique constraint and falls back to a locked increment of the winner's row instead of crashing or losing data
- 🛡️ Added `InvalidStockBalanceException` and guarded `Stock::decreaseOnHand`/`reserve`/`release` so an invalid balance is rejected at the application layer even for a caller that bypasses HTTP validation entirely
- ✅ Movement creation and balance mutation remain in the same `DB::transaction()` — a rejected mutation rolls back the `StockMovement`/`StockMovementLine` rows created earlier in the same request

## Manual Testing

1. Start the workspace: `./init-agent-workspace.sh`
2. Log in as `inventory@sushigo.com` (see Test Users table in `CLAUDE.md` for the password)
3. **Stock-out locking:** `POST /api/v1/inventory/stock-out` with a valid `inventory_location_id`/`item_variant_id`/`qty` that has existing stock — confirm it still succeeds (201) and `on_hand` decrements normally, exactly as before this change
4. **Insufficient stock still rejected:** repeat step 3 requesting more than the available `on_hand` — confirm a 400 with `Insufficient stock` (unchanged behavior)
5. **Opening balance race tolerance:** `POST /api/v1/inventory/opening-balance` twice in a row for the same `inventory_location_id` + `item_variant_id` (no existing `stock` row beforehand) — confirm both succeed (201), and only one `stock` row exists afterward with `on_hand` equal to the sum of both quantities (`SELECT * FROM stock WHERE inventory_location_id = ... AND item_variant_id = ...` should return exactly one row)
6. **DB invariant backstop:** attempt a raw update that would drive `on_hand` negative or `reserved` above `on_hand` directly against the `stock` table (e.g. via `psql`) — confirm Postgres rejects it with a CHECK constraint violation

## 🤔 Assumptions

- **Adjustments may not cross zero.** No explicit "adjustment" flow exists yet in the codebase (only `StockMovement::REASON_ADJUSTMENT` is declared as a constant, unused). The issue's own Objective requires "explicit nonnegative on-hand/reserved invariants ... for existing and future ... flows," so the most literal reading is that adjustments must never be allowed to push `on_hand` or `reserved` negative. This PR enforces that now via the DB CHECK constraints and the `StockMutationService`/`Stock` model guards, so any future adjustment flow inherits the rule automatically rather than needing to re-derive it.
- **No reservation/transfer service exists yet** (confirmed via repo search — only `StockOutService` and `OpeningBalanceService` exist today; `Stock::reserve()`/`release()` are pre-existing, unused helper methods). The issue's Objective says the invariant guarantee should hold "for existing and future receiving, reservation and transfer flows," so this PR builds the shared safety layer (`StockMutationService`, DB constraints, model guards) as the reusable foundation those flows will plug into, without adding new reservation/transfer endpoints — building those out is outside this bug fix's scope.
- **"Deterministic concurrent-request regression tests" interpreted as reproducing the race's resulting state deterministically within PHPUnit, not literal multi-connection/multi-process execution.** True overlapping transactions aren't achievable inside Laravel's `RefreshDatabase`-per-test transaction wrapper (confirmed: no such pattern exists anywhere else in this test suite). Followed the existing precedent in `tests/Feature/Media/MediaAssetDeleteTest.php::it_tolerates_deleting_an_asset_a_concurrent_actor_already_removed`, which simulates a concurrent actor's effect on shared state rather than real thread/process concurrency. This PR's tests prove the actual locking mechanism is wired in (`FOR UPDATE` via query-log assertion) and force a real `UniqueConstraintViolationException` recovery (via a genuinely pre-existing conflicting row, not a mock).

## ⚠️ Needs Human Judgment

_None yet — this section is populated automatically if an automated reviewer (Copilot/Devin) disputes the intended behavior during review._

## Test plan
- [x] PHPUnit: `php artisan test --filter="StockMutationServiceTest|StockBalanceInvariantsTest|StockOutTest|OpeningBalanceTest|StockTest"` — 45 tests passing
- [x] Full Feature suite regression check: `php artisan test --testsuite=Feature` — 1322 tests passing, 0 regressions
- [x] Linters: Pint ✅

## Workspace
`sushigo-d` — `fix/430-concurrent-stock-safety`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: code/api/app/Models/Stock.php [hunk 1]
```diff
@@ -70,25 +71,51 @@ public function increaseOnHand(float $qty): void
...
     public function decreaseOnHand(float $qty): void
     {
+        if ((float) $this->on_hand < $qty) {
+            throw new InvalidStockBalanceException(
+                "Cannot decrease on_hand below zero for stock #{$this->id}. Current: {$this->on_hand}, Requested: {$qty}"
+            );
+        }
+
         $this->decrement('on_hand', $qty);
     }
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Models/Stock.php]: Fixed in 0b90ad3 — decreaseOnHand() now also rejects a decrement that would leave on_hand below reserved, not just below zero, so it can never produce a row violating the reserved <= on_hand DB constraint. Added a regression test for this case in StockTest.

File: code/api/app/Services/Inventory/StockMutationService.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+use Illuminate\Database\UniqueConstraintViolationException;
+use Illuminate\Support\Facades\DB;
+
+/**
+ * Reusable, atomic Stock mutation service shared by every inventory flow
+ * that reads or mutates a Stock row (stock-out today; receiving, reservation
+ * and transfer flows going forward). Every method here must be called from
+ * inside an existing DB::transaction() — it does not open its own, so
+ * callers keep movement creation and balance mutation atomic together.
+ */
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Services/Inventory/StockMutationService.php]: Fixed in 0b90ad3 — updated the class docblock to call out receiveInto() as the one method that opens its own (nested/savepoint) transaction, and why, instead of implying every method requires a caller-managed transaction.

File: code/api/tests/Unit/Models/StockTest.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        $stock = $this->makeStock(3, 0);
+
+        $this->expectException(InvalidStockBalanceException::class);
+
+        $stock->decreaseOnHand(4);
+    }
+
+    #[Test]
+    public function it_reserves_available_quantity(): void
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/tests/Unit/Models/StockTest.php:112]: Fixed in 0b90ad3 — added it_rejects_decreasing_on_hand_below_reserved(), which decreases on_hand from 10 to a value that would go below reserved=8 and asserts InvalidStockBalanceException is thrown.

**Comment by pakodiazdev** [comment]
## PR #447 — Follow-up Report

### Review Comments
- Total comments: 3 (all from Copilot)
- Attended (fixed): 3
- Attended (already handled): 0
- Discarded (with justification): 0
- Pending (needs reviewer confirmation): 0

| # | Reviewer | File / Location | Comment summary | Decision | Action taken |
|---|---|---|---|---|---|
| 1 | Copilot | `code/api/app/Models/Stock.php` | `decreaseOnHand()` only guarded `on_hand >= 0`, not `reserved <= on_hand` | ✅ ATENDER | Added guard rejecting a decrement that would leave `on_hand < reserved`; regression test added in `StockTest` |
| 2 | Copilot | `code/api/app/Services/Inventory/StockMutationService.php` | Class docblock claimed no method opens its own transaction, but `receiveInto()` does | ✅ ATENDER | Docblock updated to call out `receiveInto()` as the one method with its own nested/savepoint transaction |
| 3 | Copilot | `code/api/tests/Unit/Models/StockTest.php:112` | Missing unit test for `reserved > on_hand` invariant | ✅ ATENDER | Added `it_rejects_decreasing_on_hand_below_reserved()` |

All three fixes landed in commit `0b90ad3` and are present in the branch's current squashed commit `ff8a8ac` (verified by re-reading `Stock::decreaseOnHand()`, the `StockMutationService` docblock, and `StockTest::it_rejects_decreasing_on_hand_below_reserved` on disk). Note: `0b90ad3` is no longer part of the branch's reachable history after the squash, but the SHA is still resolvable on GitHub so the reply links remain valid.

Devin: no review comments posted (DeepWiki link included in PR body per convention, no automated findings to address).

### Commit Convention
- Total commits: 1 (`ff8a8ac`, squashed via `sync-sprint-progress`)
- Compliant: 1
- Violations: 0

| Commit SHA | Subject | Opens emoji | Issue `[#NNN]` | Dash ` - ` | Closing emoji | Body bullets emoji | Verdict |
|---|---|---|---|---|---|---|---|
| `ff8a8ac` | `🐛 [#430] - Prevent concurrent stock overselling and enforce balance invariants 🔒` | ✅ 🐛 | ✅ `[#430]` | ✅ | ✅ 🔒 | ✅ all 8 bullets | ✅ Compliant |

### SonarCloud
- Quality Gate: ✅ Passed (`pakodiazdev_sushigo-api`)
- New coverage: 95.3% (≥80% required) ✅
- New bugs: 0 ✅
- New code smells / issues: 0 ✅
- Security hotspots: 0 ✅
- Duplication on new code: 0.0% ✅

### Linters
- Backend Pint: ✅ clean (951 files, 0 style violations)
- Frontend ESLint: ✅ 0 errors (28 pre-existing warnings in unrelated `src/pages/dev/-catalog-registry.tsx`, not touched by this PR)
- Frontend TypeScript: ✅ 0 errors

### Test Suite
- API Feature suite: ✅ 1325 passed, 0 regressions (316.71s)
- Webapp suite: ✅ 259 files / 3846 tests passed, 3 skipped (pre-existing skips), 0 regressions
- New code coverage: 95.3% (≥80% required) ✅

### Overall Status
✅ READY TO MERGE

### Blockers (if any)
None.

**Comment by pakodiazdev** [comment]
@codex review

File: code/api/database/migrations/2026_08_12_210000_add_balance_check_constraints_to_stock_table.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    public function up(): void
+    {
+        // NOT VALID: enforced on every write from this point forward, but
+        // doesn't scan/fail on pre-existing rows — a row already violating
+        // the invariant (e.g. from the very race this migration closes)
+        // can't block deployment, though any further write to that row is
+        // still checked and rejected until it's corrected.
+        DB::statement('ALTER TABLE stock ADD CONSTRAINT stock_on_hand_nonnegative CHECK (on_hand >= 0) NOT VALID');
+        DB::statement('ALTER TABLE stock ADD CONSTRAINT stock_reserved_nonnegative CHECK (reserved >= 0) NOT VALID');
+        DB::statement('ALTER TABLE stock ADD CONSTRAINT stock_reserved_not_exceeding_on_hand CHECK (reserved <= on_hand) NOT VALID');
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/database/migrations/2026_08_12_210000_add_balance_check_constraints_to_stock_table.php:32]: Good catch — repaired in a7feb0e. Added a single UPDATE before the constraints are created that clamps any pre-existing violation (`on_hand < 0`, `reserved < 0`, or `reserved > on_hand`) using the row's original values together, then added `VALIDATE CONSTRAINT` for all three so every existing row is confirmed compliant by the end of the migration, not just future writes.
